packages/vue-components/src/List.vue

```diff
@@ -0,0 +1,142 @@
+<template>
+  <component
+    :is="listType"
+    class="list-container"
+    :class="bulletStyle"
+  >
+    <div v-if="!parsedSlot || parsedSlot.length === 0">
+      <slot></slot>
+    </div>
+    <div
+      v-else
+    >
+      <ListItem
+        v-for="(item, index) in parsedSlot"
+        :key="index"
+        :icon="item.icon"
+      >
+        {{ item.content }}
+      </ListItem>
+    </div>
+  </component>
+</template>
+
+<script>
+import ListItem from './ListItem.vue';
+
+export default {
+  name: 'List',
+  components: {
+    ListItem,
+  },
+  provide() {
+    return {
+      list: this,
+    };
+  },
+  props: {
+    bulletStyle: {
+      type: String,
+      default: 'none',
+    },
+    bulletIcon: {
+      type: String,
+      default: '', // URL of the bullet icon image or FontAwesome class
+    },
+  },
+  computed: {
+    listType() {
+      const orderedListStyles = ['decimal', 'lower-alpha', 'upper-alpha', 'lower-roman', 'upper-roman'];
+      return orderedListStyles.includes(this.bulletStyle) ? 'ol' : 'ul';
+    },
+    parsedSlot() {
+      function adjustIndentation(code) {
+        const lines = code.split('\n');
+        let firstIndent = null;
+        const result = [];
+        let lastTopLevelLine = null;
+
+        lines.forEach((line) => {
+          if (line.trim() === '') return; // Skip empty lines
+          const lineIndent = line.match(/^\s*/)[0].length;
+          if (firstIndent === null) firstIndent = lineIndent;
+          const newLine = line.slice(firstIndent);
+          if (lineIndent > firstIndent) {
+            result[lastTopLevelLine] += `\n${newLine}`;
+          } else {
+            result.push(newLine);
+            lastTopLevelLine = result.length - 1;
+          }
+        });
+
+        return result;
+      }
+
+      const item = this.$slots.default[0];
+      if (!item.text) return undefined;
+
+      const text = item.text || '';
+      const lines = adjustIndentation(text).filter(line => line.trim() !== '');
+
+      const parseLines = (inputLines) => {
+        const items = [];
+        inputLines.forEach((line) => {
+          const matches = line.match(/^\s*{(.+?)}\s*(.+)/s);
+          const parsedItem = {
+            icon: matches ? matches[1] : null,
+            content: matches ? matches[2] : line,
+          };
+          items.push(parsedItem);
+        });
+        return items;
+      };
+
+      return parseLines(lines);
```

Comment:
Thank you for the resource provided. My apologies for not discussing the syntax prior to this.

However, given that we are integrating a new component and syntax into the current project, I believe simplifying the syntax could be beneficial. Specifically, we have tags like <List> that function as a domain, ensuring that all its children are elements of the list.

For example, the existing structure is:
```
<List>
  {fas fa-check text-danger} first item
  second item
  {fas fa-check text-success} third item
  {/images/deer.jpg} fourth item
  fifth item
</List>
```
```
<icon-list>
* {.icon=":fas-tick:"} first item
* second item
  * {.icon=":fas-x:" .class="text-danger"} third item
  * fourth item
</icon-list>
```
The list symbols *, -, and 1. are currently used for the original markdown parser to identify the list component. However, given that we're already using the tag <List>, it seems redundant to keep these symbols.

1) Apologies as I'm still acquainting myself with the project; I'm uncertain about the exact meaning of 'backend/frontend' in your context. If by 'backend' you mean the processing of HTML code server-side and 'frontend' refers to the DOM manipulation performed client-side via JavaScript:
    - Given this understanding, I believe that employing either approach wouldn't significantly impact user experience as our current project involves minimal DOM manipulation. However, relying on client-side rendering might not be advantageous for search engines or browser, given that the components we're adding don't comply with the HTML5.

2) If the project only involves DOM manipulation, the necessity of Vue components could be questioned. However, I find that utilizing Vue streamlines the implementation of new components. The advantages of using Vue components include:
    - Effortless component location by tag.
    - The ability to apply different rule sets to distinct tags.
    - Convenient code referencing or locating for a specific feature within the Vue component.

Reply:
>Let's try to get an ok on the syntax first too from @damithc and anyone else before going further to avoid wasted work.
We could continue that here https://github.com/MarkBind/markbind/issues/2272

Yup 👍 

>The list symbols *, -, and 1. are currently used for the original markdown parser to identify the list component. However, given that we're already using the tag , it seems redundant to keep these symbols.

I think they are still needed because you can't assume every line is a separate item. Some items might span multiple lines. Also, visual consistency is good to have (i.e., a list in code should still _look like_ a list).

>However, given that we are integrating a new component and syntax into the current project, I believe simplifying the syntax could be beneficial. Specifically, we have tags like that act as a domain, ensuring that all its children are elements of the list.

It is also important to maintain syntax consistency over the entire project.

Reply:
> I believe simplifying the syntax could be beneficial.

we use this plugin too https://www.npmjs.com/package/markdown-it-attrs which may give you some ideas to make the syntax even more user friendly (or not) and easily integratable with existing markdown lists.

> If by 'backend' you mean the processing of HTML code server-side and 'frontend' refers to the DOM manipulation performed client-side via JavaScript:

Yes, sorry, I couldn't find a better word 😅 

> Given this understanding, I believe that employing either approach wouldn't significantly impact user experience as our current project involves minimal DOM manipulation. However, relying on client-side 

Agreed probably not significantly, unless on slower devices or larger pages. But if it can be easily avoided and implemented we should aim to eliminate that possibility entirely for consistent ux.

(actually the same point on the comment below on FOUC being caused by manipulating the DOM into its "final form")

> If the project only involves DOM manipulation, the necessity of Vue components could be questioned. 

👍 You can try both out and see how it goes.

Reply:
> It is also important to maintain syntax consistency over the entire project.

Adding on there are a lot of other obscure rules in the spec as well that are an element of surprise for the author if not followed.

E.g. This won't work

```
<div>
- my
- list
- items
</div>
```

But this will


```
<div>

- my
- list
- items
</div>
```

Reply:
Thank you for pointing this out. This indeed is a worthwhile question to consider.

Currently, Markbind supports syntax based on the given spec:
```
<div>

- my
- list
- items
</div>
```

However, I believe it would make more sense to adopt the following format:

```
<div>
- my
- list
- items
</div>
```

Which one would you refer?

Reply:
I agree with @damithc on consistency (hence the former), it is a golden rule of UX design to adopt the path of least surprise.

Please note I've just quoted one rule of many too. The more important point is that we shouldn't be attempting to reinevent the markdown parsing wheel, it is really difficult to implement correctly.

https://github.com/markdown-it/markdown-it
https://github.com/markedjs/marked

Reply:
> we use this plugin too https://www.npmjs.com/package/markdown-it-attrs which may give you some ideas to make the syntax even more user friendly (or not) and easily integratable with existing markdown lists.

This is what I have in mind currently but feel free to discuss and improve on it with @damithc.

```
- something { iconOrMediaOverride=":octicon-...:" }
- like this

{ iconOrMediaListWideDefault="foobar.png" }
```

What @damithc is suggesting in the issue comment is a small variant

```
<medialist iconOrMediaListWideDefault="foobar.png">

- something { iconOrMediaOverride=":octicon-...:" }
- like this

</medialist>
```

I'm tending toward the former as its consistent with how we support code block features using the `{...}` attributes as well and don't think we need the flexibility of a HTML component (if we do, we can always add on later).

Reply:
> This is what I have in mind currently but feel free to discuss and improve on it with @damithc.
> 
> ```
> - something { iconOrMediaOverride=":octicon-...:" }
> - like this
> 
> { iconOrMediaListWideDefault="foobar.png" }
> ```
> 

I'm OK to have the icon override at the end of an item/list for the sake of consistency with the rest of the syntax (perhaps easier to process too), rather than at the start.
Perhaps we can make the attribute name shorter?

@SPWwj let's discuss the syntax at https://github.com/MarkBind/markbind/issues/2272